src/Workspaces/Core/Portable/CodeRefactorings/SyntaxEditorBasedCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using FixAllScope = Microsoft.CodeAnalysis.CodeFixes.FixAllScope;
+
+namespace Microsoft.CodeAnalysis.CodeRefactorings
+{
+    internal abstract partial class SyntaxEditorBasedCodeRefactoringProvider : CodeRefactoringProvider
+    {
+        protected static readonly ImmutableArray<FixAllScope> DefaultFixAllScopes = ImmutableArray.Create(FixAllScope.Document,
+            FixAllScope.Project, FixAllScope.Solution);
+        protected static readonly ImmutableArray<FixAllScope> AllFixAllScopes = ImmutableArray.Create(FixAllScope.Document,
+            FixAllScope.Project, FixAllScope.Solution, FixAllScope.ContainingType, FixAllScope.ContainingMember);
+
+        protected abstract ImmutableArray<FixAllScope> SupportedFixAllScopes { get; }
+
+        internal sealed override FixAllProvider? GetFixAllProvider()
+        {
+            if (SupportedFixAllScopes.IsEmpty)
+                return null;
+
+            return FixAllProvider.Create(
+                async (fixAllContext, document, fixAllSpans) =>
+                {
+                    return await this.FixAllAsync(document, fixAllSpans, fixAllContext.CancellationToken).ConfigureAwait(false);
+                },
```

Comment:
```suggestion
                (fixAllContext, document, fixAllSpans) => this.FixAllAsync(document, fixAllSpans, fixAllContext.CancellationToken),
```

Reply:
Nullability of return type of `this.FixAllAsync` and this lambda are different (`Task<Document?>` vs `Task<Document>`), so the existing syntax helps.